File: week18/week18-2.cpp
//week18-2.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int M= matrix.size(),N=matrix[0].size(); //長寬
        int i=0,j=0; //開始點
        int di[4] = {0,1,0,-1};
        int dj[4] = {1,0,-1,0};
        int d=0; //右 下 左 上
        vector<int> ans;
        while( ans.size() < M*N ){
            ans.push_back( matrix[i][j]);
            matrix[i][j] = 999; //999代表走過了
            int i2=i+di[d], j2=j+dj[d];
            if(i2>=M || j2>=N || i2<0 || j2<0 || matrix[i2][j2]==999){
                d = (d+1)%4;
            }
            i=i+di[d];
            j=j+dj[d];
        }
        return ans;
    }
};

File: week18/week18-2.py
#week18-2.py
#LeetCode 54. Spiral Matrix
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        M,N= len(matrix), len(matrix[0])
        i,j = 0,0
        di = [0,1,0,-1] #每次i移動的量
        dj = [1,0,-1,0] #每次j移動的量
        d=0 #移動的方向,對應 di[d] dj[d]
        ans=[]
        while len(ans) < M*N: #還沒收齊12個答案
            ans.append( matrix[i][j] )
            matrix[i][j] = 999 #用999代表走過了
            i2, j2 = i+di[d], j+dj[d] #下一格
            if i2>=M or j2>=N or i2<0 or j2<0 or matrix[i2][j2]==999: #撞到邊界 or 走過
                d= (d+1) %4 #轉彎
                i2,j2 = i+di[d], j+dj[d]
                #print(i,j,"turn")
            i,j =i2,j2
        return ans